### ISSUE TITLE
House Sales in King County, USA.

### ISSUE THREAD

**Issue by DhruvRajSinghRathore007** [DhruvRajSinghRathore007/dhruv#4]
Module 1: Importing Data Sets

First, let's load the dataset.

Python
import pandas as pd
import numpy as np

filepath = 'https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-DA0101EN-SkillsNetwork/labs/FinalModule_Coursera/data/kc_house_data_NaN.csv'
df = pd.read_csv(filepath)
Question 1: Display the data types of each column.

Here is the code to display the data types of each column in the dataframe.

Code:

Python
print(df.dtypes)
Output:

Unnamed: 0         int64
id                 int64
date              object
price            float64
bedrooms         float64
bathrooms        float64
sqft_living        int64
sqft_lot           int64
floors           float64
waterfront         int64
view               int64
condition          int64
grade              int64
sqft_above         int64
sqft_basement      int64
yr_built           int64
yr_renovated       int64
zipcode            int64
lat              float64
long             float64
sqft_living15      int64
sqft_lot15         int64
dtype: object
Module 2: Data Wrangling

Question 2: Drop the columns "id" and "Unnamed: 0" and then obtain a statistical summary of the data.

This task involves dropping the specified columns and then using the describe() method.

Code:

Python
df.drop(["id", "Unnamed: 0"], axis=1, inplace=True)
print(df.describe())
Output:

              price      bedrooms     bathrooms   sqft_living      sqft_lot  \
count  2.161300e+04  21600.000000  21603.000000  21613.000000  2.161300e+04   
mean   5.400881e+05      3.372870      2.115736   2079.899736  1.510697e+04   
std    3.671272e+05      0.926657      0.768996    918.440897  4.142051e+04   
min    7.500000e+04      1.000000      0.500000    290.000000  5.200000e+02   
25%    3.219500e+05      3.000000      1.750000   1427.000000  5.040000e+03   
50%    4.500000e+05      3.000000      2.250000   1910.000000  7.618000e+03   
75%    6.450000e+05      4.000000      2.500000   2550.000000  1.068800e+04   
max    7.700000e+06      33.000000      8.000000  13540.000000  1.651359e+06   

            floors    waterfront          view     condition         grade  \
count  21613.000000  21613.000000  21613.000000  21613.000000  21613.000000   
mean       1.494309      0.007542      0.234303      3.409430      7.656873   
std        0.539989      0.086517      0.766318      0.650743      1.175459   
min        1.000000      0.000000      0.000000      1.000000      1.000000   
25%        1.000000      0.000000      0.000000      3.000000      7.000000   
50%        1.500000      0.000000      0.000000      3.000000      7.000000   
75%        2.000000      0.000000      0.000000      4.000000      8.000000   
max        3.500000      1.000000      4.000000      5.000000     13.000000   

         sqft_above  sqft_basement      yr_built  yr_renovated       zipcode  \
count  21613.000000   21613.000000  21613.000000  21613.000000  21613.000000   
mean    1788.390691     291.509045   1971.005136    84.402258  98077.939805   
std      828.090978     442.575043     29.373411    401.679246     53.505026   
min      290.000000       0.000000   1900.000000      0.000000  98001.000000   
25%     1190.000000       0.000000   1951.000000      0.000000  98033.000000   
50%     1560.000000       0.000000   1975.000000      0.000000  98065.000000   
75%     2210.000000     560.000000   1997.000000      0.000000  98118.000000   
max     9410.000000    4820.000000   2015.000000   2015.000000  98199.000000   

               lat          long  sqft_living15    sqft_lot15  
count  21613.000000  21613.000000   21613.000000  21613.000000  
mean      47.560053    -122.213896    1986.552492  12768.455652  
std        0.138564       0.140828     685.391304  27304.179631  
min       47.155900    -122.519000     399.000000    651.000000  
25%       47.471000    -122.328000    1490.000000   5100.000000  
50%       47.571800    -122.230000    1840.000000   7620.000000  
75%       47.678000    -122.125000    2360.000000  10083.000000  
max       47.777600    -121.315000    6210.000000  871200.000000  
Before proceeding, let's also handle the missing values in 'bedrooms' and 'bathrooms' as indicated in the notebook.

Python
mean_bedrooms = df['bedrooms'].mean()
df['bedrooms'].replace(np.nan, mean_bedrooms, inplace=True)

mean_bathrooms = df['bathrooms'].mean()
df['bathrooms'].replace(np.nan, mean_bathrooms, inplace=True)
Module 3: Exploratory Data Analysis

Question 3: Count the number of houses with unique floor values and convert it to a data frame.

Code:

Python
floor_counts = df['floors'].value_counts().to_frame()
print(floor_counts)
Output:

        floors
1.0    10680
2.0     8241
1.5     1910
3.0      613
2.5      161
3.5        8
Question 4: Use a boxplot to determine if houses with a waterfront view have more price outliers.

Code:

Python
import seaborn as sns
import matplotlib.pyplot as plt

plt.figure(figsize=(10, 6))
sns.boxplot(x='waterfront', y='price', data=df)
plt.title('Price Distribution by Waterfront View')
plt.show()
Output:

From the boxplot, we can observe that houses with a waterfront view (waterfront=1) have a higher median price and a wider distribution of prices, with several outliers at the higher end. Houses without a waterfront view (waterfront=0) also have many outliers, but their prices are generally lower.

Question 5: Use a regression plot to determine if sqft_above is positively or negatively correlated with price.

Code:

Python
plt.figure(figsize=(10, 6))
sns.regplot(x='sqft_above', y='price', data=df)
plt.title('Square Feet Above vs. Price')
plt.show()
Output:

The regression plot shows a positive slope, indicating that sqft_above and price are positively correlated. As the square footage above ground increases, the price of the house tends to increase.

Module 4: Model Development

Question 6: Fit a linear regression model to predict price using sqft_living and calculate the R-squared.

Code:

Python
from sklearn.linear_model import LinearRegression

X = df[['sqft_living']]
Y = df['price']

lm = LinearRegression()
lm.fit(X, Y)
r2_sqft_living = lm.score(X, Y)

print(f"The R-squared value is: {r2_sqft_living}")
Output:

The R-squared value is: 0.4928532179037931
Question 7: Fit a multiple linear regression model to predict price using a list of features and calculate the R-squared.

Code:

Python
features = ["floors", "waterfront", "lat", "bedrooms", "sqft_basement", "view", "bathrooms", "sqft_living15", "sqft_above", "grade", "sqft_living"]
X_multi = df[features]
Y = df['price']

lm_multi = LinearRegression()
lm_multi.fit(X_multi, Y)
r2_multi = lm_multi.score(X_multi, Y)

print(f"The R-squared value for the multiple linear regression model is: {r2_multi}")
Output:

The R-squared value for the multiple linear regression model is: 0.6576951666037503
Module 5: Model Evaluation and Refinement

Question 8: Create a pipeline to scale data, perform a polynomial transform, and fit a linear regression model. Calculate the R-squared.

Code:

Python
from sklearn.pipeline import Pipeline
from sklearn.preprocessing import StandardScaler, PolynomialFeatures

Input = [('scale', StandardScaler()), ('polynomial', PolynomialFeatures(degree=2, include_bias=False)), ('model', LinearRegression())]
pipe = Pipeline(Input)

X_pipe = df[features]
Y = df['price']

pipe.fit(X_pipe, Y)
r2_pipe = pipe.score(X_pipe, Y)

print(f"The R-squared value for the pipeline model is: {r2_pipe}")
Output:

The R-squared value for the pipeline model is: 0.7513410756335191
Question 9: Create and fit a Ridge regression object, and calculate the R-squared using the test data.

First, we need to split the data into training and testing sets.

Python
from sklearn.model_selection import train_test_split

features = ["floors", "waterfront", "lat", "bedrooms", "sqft_basement", "view", "bathrooms", "sqft_living15", "sqft_above", "grade", "sqft_living"]
X = df[features]
Y = df['price']

x_train, x_test, y_train, y_test = train_test_split(X, Y, test_size=0.15, random_state=1)
Now, we can create and fit the Ridge regression model.

Code:

Python
from sklearn.linear_model import Ridge

ridge_model = Ridge(alpha=0.1)
ridge_model.fit(x_train, y_train)
r2_ridge = ridge_model.score(x_test, y_test)

print(f"The R-squared value for the Ridge regression model on the test data is: {r2_ridge}")
Output:

The R-squared value for the Ridge regression model on the test data is: 0.6478759163939115
Question 10: Perform a second-order polynomial transform and fit a Ridge regression model. Calculate the R-squared using the test data.

Code:

Python
poly = PolynomialFeatures(degree=2)
x_train_poly = poly.fit_transform(x_train)
x_test_poly = poly.transform(x_test)

ridge_model_poly = Ridge(alpha=0.1)
ridge_model_poly.fit(x_train_poly, y_train)
r2_ridge_poly = ridge_model_poly.score(x_test_poly, y_test)

print(f"The R-squared value for the Ridge regression model with polynomial features on the test data is: {r2_ridge_poly}")
Output:

The R-squared value for the Ridge regression model with polynomial features on the test data is: 0.7002734680888975
